text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.
+
+The `Receiver` trait is simple and only requires to specify the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(self: *const Self);
+    fn by_mut_ptr(self: *mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward).
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized, T: ?Sized> Receiver for P
+where
+    P: Deref<Target = T>,
+{
+    type Target = T;
+}
+```
+
+(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)
+
+It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.
+
+## Compiler changes
+
+The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
```

Comment:
I think this is subtler than it looks: if `x: &Arc<T>` and I call `x.foo()`, the current algorithm will (simplified):
- look for a `foo(&self)` method on `Arc`
- if not found, check if `Arc` implements `Deref`
- if yes, recursively look for a `foo(&self)` method on `<Arc<T> as Deref>::Target` (i.e. `T`)
- if not found, check if `T` implements `Deref`
- etc.

Your proposed algorithm would:
- look for a `foo(&self)` method on `Arc`
- if not found, check if `Arc` implements `Receiver`
- if yes, recursively (?) look for a `foo(&Arc<Self>)` method on `<Arc<T> as Receiver>::Target`
- if not found, what do we do next?

The algorithm needs at some point to look for a `foo(&self)` on `T`. I think this requires special-casing `Deref` still

Reply:
It gets fun with nesting: say `x: &Arc<Box<T>>`. We need to check:
- `foo(&self)` on `Arc<Box<T>>`
- `foo(&Arc<Self>)` on `Box<T>`
- `foo(&Arc<Box<Self>>)` on `T`
- `foo(&self)` on `Box<T>`
- `foo(&Box<Self>)` on `T`
- `foo(&self)` on `T`

We'd need to decide on which should take priority, I think the order I listed them in is sensible. An algorithm for this would look like "try more and more layers of `Receiver` until we can't anymore, then take one layer of `Deref`, then start again.

Reply:
Ah, I had missed [the section](https://github.com/adetaylor/rfcs/blob/arbitrary-self-types-v2/text/3519-arbitrary-self-types-v2.md#explore-both-receiver-and-deref-chains-while-identifying-method-candidates) where you talk about this, sorry. I still don't quite understand how there's a way to avoid `Deref` entirely, even with a blanket `impl<P: Deref> Receiver for P`

Reply:
We went backwards and forwards on how much detail to put in this section of the RFC! In the end we decided to just state the specific small change we're making, instead of trying to describe the current algorithm in lots of detail.

The current unstable `arbtrary_self_types` complains of an ambiguity in the [case of your example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6eeafbe7e26a7d02f870c14f9d2ebf71).

```
error[E0034]: multiple applicable items in scope
  --> src/main.rs:54:7
   |
54 |     x.foo();
   |       ^^^ multiple `foo` found
   |
note: candidate #1 is defined in an impl for the type `Arcish<Boxish<T>>`
  --> src/main.rs:24:5
   |
24 |     fn foo(&self) {
   |     ^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `Boxish<T>`
  --> src/main.rs:30:5
   |
30 |     fn foo(self: &Arcish<Self>) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #3 is defined in an impl for the type `Concrete`
  --> src/main.rs:39:5
   |
39 |     fn foo(self: &Arcish<Boxish<Self>>) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

That probably needs to change - a [comment thread above notes that this can cause inter-crate compatibility breakages](https://github.com/rust-lang/rfcs/pull/3519/files#r1379664429).

Buuuutt... I don't _think_ there's a need for `Deref` here. I think with a blanket implementation `impl<P: Deref> Receiver for P` the method candidate selection algorithm should never need to consider `Deref` explicitly. Can you explain more about why you think it might?

Reply:
Hmm, I thought my detailed steps above made my point clear ><. Let me try again.

I seem to understand that there's a fundamental difference between what `Deref` and `Receiver` mean in this context. `Receiver` on `P<T>` means "look for methods like `foo(P<Self>)` on `T`. `Deref` on `P<T>` means "look for methods like `foo(&self)` on `T`. `Receiver` only ever allows you to look for methods that take `P<T>`, the `P` must be mentioned. Contrast with `Deref`, which removes one layer of pointer.

Assume the resolution algorithm only uses `Receiver`, and we have the blanket `Deref` impl. Assume I have `x: Arc<T>` and I call `x.foo()`. What we clearly do want is to call `foo(Arc<Self>)` if there is one. I believe this is what the `Receiver` impl means. Yet if there is no such method, we also want to call `foo(&self)` on `T` if there is one! And I don't see how `Receiver` on its own allows this.

Reply:
OK - I see what you mean, and I can see the part of the description which I need to clarify. The proposed algorithm won't use the `Deref` trait at all, but it _will_ retain the current logic which searches for method candidates not only on `T` but also `&T` and `&mut T`. I'll write something about that.

Reply:
I clarified a little - hopefully that helps! Please let me know if not.

Reply:
Small comment: You marked the question resolved so I almost missed that you had replied to me. I find it a good convention to let whoever asked the question mark it as "resolved", unless it's a really obvious comment like a typo